diag/diagnostics.go

```diff
@@ -0,0 +1,97 @@
+package diag
+
+import (
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// Diagnostics represents a collection of diagnostics.
+//
+// While this collection is ordered, the order is not guaranteed as reliable
+// or consistent.
+type Diagnostics []Diagnostic
+
+// AddAttributeError adds a generic attribute error diagnostic to the collection.
+func (diags *Diagnostics) AddAttributeError(path *tftypes.AttributePath, summary string, detail string) {
+	diags.Append(NewAttributeErrorDiagnostic(path, summary, detail))
+}
+
+// AddAttributeWarning adds a generic attribute warning diagnostic to the collection.
+func (diags *Diagnostics) AddAttributeWarning(path *tftypes.AttributePath, summary string, detail string) {
+	diags.Append(NewAttributeWarningDiagnostic(path, summary, detail))
+}
+
+// AddError adds a generic error diagnostic to the collection.
+func (diags *Diagnostics) AddError(summary string, detail string) {
+	diags.Append(NewErrorDiagnostic(summary, detail))
+}
+
+// AddWarning adds a generic warning diagnostic to the collection.
+func (diags *Diagnostics) AddWarning(summary string, detail string) {
+	diags.Append(NewWarningDiagnostic(summary, detail))
+}
+
+// Append adds non-empty and non-duplicate diagnostics to the collection.
+func (diags *Diagnostics) Append(in ...Diagnostic) {
+	for _, diag := range in {
+		if diag == nil {
+			continue
+		}
+
+		if diags.Contains(diag) {
+			continue
+		}
+
+		if diags == nil {
+			*diags = Diagnostics{diag}
+		} else {
+			*diags = append(*diags, diag)
+		}
+	}
+}
+
+// Contains returns true if the collection contains an equal Diagnostic.
+func (diags Diagnostics) Contains(in Diagnostic) bool {
+	for _, diag := range diags {
+		if diag.Equal(in) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// HasError returns true if the collection has an error severity Diagnostic.
+func (diags Diagnostics) HasError() bool {
+	for _, diag := range diags {
+		if diag.Severity() == SeverityError {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ToTfprotov6Diagnostics converts the diagnostics into the tfprotov6 collection type.
+//
+// Usage of this method outside the framework is not supported nor considered
+// for backwards compatibility promises.
+func (diags Diagnostics) ToTfprotov6Diagnostics() []*tfprotov6.Diagnostic {
```

Comment:
two things:

1) This doesn't appear to require any privileged access to the `diag` package or benefit from being a method on the `Diagnostics` type. Is there a reason we shouldn't just make it a function in `tfsdk`?
2) I think `Tfprotov6` in this may be unnecessarily verbose. `ToProto6Diagnostics` is probably sufficient; we know which protocol we're talking about, and we're unlikely to be talking about any other protocol's diagnostics except Terraform's.

Reply:
> This doesn't appear to require any privileged access to the diag package or benefit from being a method on the Diagnostics type. Is there a reason we shouldn't just make it a function in tfsdk?

It does keep the diagnostics implementation altogether.

Reply:
Hm. Is that worth having an exported method we don't want people to call? 🤔 (I don't know the answer, I'm wondering out loud.)

Reply:
Another option here would be to move that exported function into a separate `internal` package, maybe along the lines of `diag/internal/convert`. I'd be more apt to do this so at least the code is tangential the other diagnostics code. I'd be worried about import cycles if we tried to setup something more generic like `internal/toproto6`.

@kmoe do you have any opinions?

Reply:
I'd still prefer this to be unexported as we don't expect provider developers to use it, and I'm worried about what they'll do with this unintended compatibility interface, but I'm not gonna hold this up over it.